packages/expo-modules-core/ios/Swift/Logging/PersistentFileLog.swift

```diff
@@ -0,0 +1,152 @@
+// Copyright 2022-present 650 Industries. All rights reserved.
+
+import Foundation
+
+public typealias PersistentFileLogFilter = (String) -> Bool
+public typealias PersistentFileLogCompletionHandler = (Error?) -> Void
+
+/**
+ * A thread-safe class for reading and writing line-separated strings to a flat file.
+ * The main use case is for logging specific errors or events, and ensuring that the
+ * logs persist across application crashes and restarts (for example, OSLogReader can
+ * only read system logs for the current process, and cannot access anything logged
+ * before the current process started).
+ *
+ * All write access to the file goes through asynchronous public methods managed by a
+ * serial dispatch queue.
+ *
+ * The dispatch queue is global, to ensure that multiple instances accessing the same
+ * file will have thread-safe access.
+ *
+ * The only operations supported are
+ * - Read the file (synchronous)
+ * - Append one or more entries to the file
+ * - Filter the file (only retain entries that pass the filter check)
+ * - Clear the file (remove all entries)
+ *
+ */
+public class PersistentFileLog {
+  private static let EXPO_MODULES_CORE_LOG_QUEUE_LABEL = "dev.expo.modules.core.logging"
+  private static let serialQueue = DispatchQueue(label: EXPO_MODULES_CORE_LOG_QUEUE_LABEL)
+
+  private let category: String
+  private let filePath: String
+
+  public init(category: String) {
+    self.category = category
+    let fileName = "\(PersistentFileLog.EXPO_MODULES_CORE_LOG_QUEUE_LABEL).\(category).txt"
+    // Execution aborts if no documents directory
+    self.filePath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!.appendingPathComponent(fileName).path
+  }
+  /**
+   Read entries from log file
+   */
+  public func readEntries() -> [String] {
```

Comment:
(applies on Android also) What happens if one thread tries to read from the file at the same time as another one is appending a new log entry? I am guessing we would end up reading a partial entry and simply discard it. I think this would be ok but want to make sure it's an intentional choice. I could see an argument for having reads happen asynchronously on the `serialQueue` as well.

Reply:
Correct, and this is one reason why the code that parses log lines will just return null if a line is truncated.